amethyst_animation/src/resources.rs

```diff
@@ -0,0 +1,193 @@
+use std::time::{Duration, Instant};
+
+use amethyst_assets::{Asset, Handle, Result};
+use fnv::FnvHashMap;
+use specs::{Component, DenseVecStorage, Entity};
+
+use interpolation::InterpolationType;
+
+/// The actual animation data for a single attribute
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum AnimationOutput {
+    /// Translation is a 3d vector
+    Translation(Vec<[f32; 3]>),
+    /// Rotation is a quaternion
+    Rotation(Vec<[f32; 4]>),
+    /// Scale is a 3d vector
+    Scale(Vec<[f32; 3]>),
+}
+
+/// Sampler defines a single animation for a single attribute of the `LocalTransform` of the entity
+/// it is attached to.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Sampler {
+    /// Time of key frames
+    pub input: Vec<f32>,
+    /// Actual output data to interpolate
+    pub output: AnimationOutput,
+    /// How should interpolation be done
+    pub ty: InterpolationType,
+}
+
+impl Asset for Sampler {
+    type Data = Self;
+    type HandleStorage = DenseVecStorage<Handle<Self>>;
+}
+
+impl Into<Result<Sampler>> for Sampler {
+    fn into(self) -> Result<Sampler> {
+        Ok(self)
+    }
+}
+
+/// Defines the hierarchy of nodes that a single animation can control.
+/// Attach to the root entity that an animation can be defined for.
+/// Only required for animations which targets more than a single node.
+#[derive(Debug, Clone)]
+pub struct AnimationHierarchy {
+    pub nodes: FnvHashMap<usize, Entity>,
+}
+
+impl Component for AnimationHierarchy {
+    type Storage = DenseVecStorage<Self>;
+}
+
+/// Defines a single animation.
+/// Defines relationships between the node index in `AnimationHierarchy` and a `Sampler` handle.
+/// If the animation only targets a single node index, `AnimationHierarchy` is not required.
+#[derive(Clone, Debug)]
+pub struct Animation {
+    /// node index -> sampler handle
+    pub nodes: Vec<(usize, Handle<Sampler>)>,
```

Comment:
No, because you can play animation on many different entities when you use prefabs.

Reply:
You want to be able to share animations between many families of entities.

Reply:
Can you describe use case?

Reply:
A quite common scenario for indie game development is that you have a single bone structure for many different units (think human, vulkan, klingon etc), and all units that share bone structure can share animations. This saves on time and cost of developing the game, because you don't have to produce as many animations.

So, animations are loaded once and then potentially shared between many different types of units/entities whatever. 

The second issue with the suggestion you have is one of forward lookup. Given an animation and entity to play that animation on, how do i find which child entities to add samplers to ? If the only link we have is that the child entities have a reference to the entity of the prefab they were created from, you would have to look through the entire `Parent` storage to find the entitites that should be animated, which would be far to expensive imho. The other solution would be to have the `AnimationHierarchy` contain Entity -> Entity mappings. I went with the "node index -> Entity" solution, because it makes animation asset loading much, much easier.

Reply:
Animations can also target only a part of a hierarchy, not necessarily the entire hierarchy. Example: Sword swing will probably only affect arms and upper torso.

Reply:
Regardless, this is not something that can be done within the scope of this PR, because it needs prefab loading first.

Reply:
> Example: Sword swing will probably only affect arms and upper torso.

But this system animate `LocalTransform` of entities. Will it share some parts with skeletal animation system?

Reply:
Skeletal animation is this system together with vertex skinning.